1.x shared element transition and movable content integration

KovalevAndrey opened the pull request:
## Description

Share element transition integration into Appyx. 

1. Clients need to use `withSharedElementTransition: Boolean = true` on `Children` layout when drawing children.
2. Use `sharedElement` Modifier In Node's UI implementation and provide a key that matches the appropriate element's key in the other Node's UI. 

Added support for movableContent. Now we can "move one part of the composable tree" from one child node to another without losing its state. We can only move content between the Node that is transitioning to invisible state to the Node that is transitioning to visible state because movable content is meant to be composed only once at any time. For instance: `NodeA` is currently shown in a `BackStack` and we push `NodeB` to `BackStack`. In this case we can move content from `NodeA` to `NodeB`. To move the content we need to use `localMovableContentWithTargetVisibility` with the correct key in each child Node.

**Do not merge unit the [crash](https://issuetracker.google.com/issues/333679741) is released in alpha08** 

https://github.com/bumble-tech/appyx/assets/5773436/ce5a3def-68b9-4bfc-ac05-6de14080e7e1


## Checklist

- [x] I've updated `CHANGELOG.md` if required.
- [x] I've updated the documentation if required.

KovalevAndrey committed:
Shared element transition integration

KovalevAndrey committed:
Update CHANGELOG.md

KovalevAndrey committed:
Replace SharedTransitionLayout with Box where no sharedElement is supported

KovalevAndrey committed:
Update crashing lint

manuelvicnt commented on libraries/core/src/main/kotlin/com/bumble/appyx/core/composable/Children.kt:
```diff
@@ -50,18 +58,42 @@ inline fun <reified NavTarget : Any, State> ParentNode<NavTarget>.Children(
-    Box(modifier = modifier
-        .onSizeChanged {
-            transitionBounds = it
+    if (withSharedElementTransition) {
```
I assume `withSharedElementTransition` won't change dynamically that often but how the API is set up, it _could_. Should we wrap the repeated code (i.e. `CompositionLocalProvider(...) { block(...) }`) with the `moveableContentOf` API to 1) avoid repeating this code and 2) save state if `withSharedElementTransition` were to change?

KovalevAndrey replied:
I can extract repeatable parts to a separate composable but it doesn't make sense to you `moveableContentOf` the layout either supports shared element transition or not. I can't come up with a case when we need dynamically change it. WDYT?

manuelvicnt replied:
The use case could be that an app allows the user to enable/disable animations on the fly. It's definitely an edge case and something that wouldn't happen very often. But given this is an open source Architecture library and this affects the public API, I guess people could use it for _anything_. If I change `withSharedElementTransition` on the fly, I wouldn't want to lose state.

manuelvicnt commented on libraries/core/src/main/kotlin/com/bumble/appyx/core/composable/Children.kt:
```diff
@@ -50,18 +58,42 @@ inline fun <reified NavTarget : Any, State> ParentNode<NavTarget>.Children(
+            }
+        ) {
+            CompositionLocalProvider(
+                LocalSharedElementScope provides this
```
Should we document why we're providing `this` vs `null`? is it required?

KovalevAndrey replied:
I can document, yes. Null is required because if we don't pass it `LocalSharedElementScope` can still be consumed  on grand children level even if their parent doesn't support SharedElement transition

manuelvicnt commented on libraries/core/src/main/kotlin/com/bumble/appyx/core/navigation/NavModelAdapter.kt:
```diff
@@ -9,6 +9,7 @@ interface NavModelAdapter<NavTarget, State> {
     val screenState: StateFlow<ScreenState<NavTarget, out State>>
 
     data class ScreenState<NavTarget, State>(
+        val targetStateVisible: NavElements<NavTarget, out State> = emptyList(),
```
Can we add more info about what this `targetStateVisible` represents? It's a bit out of context in the `ScreenState` class. 

`onScreen` and `offScreen` makes sense but not `targetStateVisible`. How's that state supposed to mutate/evolve over time?

KovalevAndrey replied:
yeah, it conceptually belongs to the `onScreen` element. We need to know which elements from the `onScreen` list are transitioning to `visibleTargetState` but it will break the API if I make a wrapper class

manuelvicnt replied:
Gotcha! Can we be more explicit then? `onScreenVisibleTargetState` perhaps?

manuelvicnt commented on libraries/core/src/main/kotlin/com/bumble/appyx/core/composable/Children.kt:
```diff
@@ -140,30 +175,32 @@ class ChildrenTransitionScope<T : Any, S>(
+            .forEach { navElement ->
+                key(navElement.key.id) {
+                    CompositionLocalProvider(
+                        LocalNodeTargetVisibility provides
```
Are there any tests covering this? Should we add some of them? Feels like the `ScreenState` and a few other things could benefit from them? Thoughts?

KovalevAndrey replied:
I can add a UI test for that

KovalevAndrey replied:
added

KovalevAndrey committed:
Movable content support and test for targetVisibilityState

KovalevAndrey committed:
Update detekt-baseline.xml

KovalevAndrey committed:
Rename button to Shared element with movable content

KovalevAndrey committed:
Update emulator API

KovalevAndrey committed:
Add documentation for movable content and shared element transition

KovalevAndrey committed:
Update CHANGELOG.md

KovalevAndrey committed:
Add docs and uninstall test app

KovalevAndrey committed:
Uninstall RIBs test apk

KovalevAndrey committed:
Rename package and uninstall navigation compose test apk

KovalevAndrey committed:
Fix compilation error

KovalevAndrey committed:
Fix compilation error

KovalevAndrey committed:
Merge branch '1.x-shared-element-prototype' of github.com:KovalevAndrey/appyx into 1.x-shared-element-prototype

KovalevAndrey committed:
Add proguard rules release  to sandbox app

KovalevAndrey committed:
Fix testing package name

KovalevAndrey committed:
Update compose to alpha08

KovalevAndrey committed:
Temporary remove uninstalling com.bumble.appyx.sample.navigation.compose.test

KovalevAndrey committed:
Update to alpha08

manuelvicnt commented on documentation/ui/transitions.md:
```diff
@@ -12,9 +12,133 @@ Below you can find the different options how to visualise `NavModel` state chang
+
+To support shared element transition between two Child Nodes you need:
+
+1. Use sharedElementModifier with the same key on the composable you want to connect.
```
```suggestion
1. Use the `sharedElement` Modifier with the same key on the composable you want to connect.
```

manuelvicnt commented on documentation/ui/transitions.md:
```diff
@@ -12,9 +12,133 @@ Below you can find the different options how to visualise `NavModel` state chang
+      ) { /** ... */ }
+   }
+}
+class NodeOne(
```
```suggestion

class NodeTwo(
```

manuelvicnt commented on documentation/ui/transitions.md:
```diff
@@ -12,9 +12,133 @@ Below you can find the different options how to visualise `NavModel` state chang
+         navModel = backStack,
+         // or no transitionHandler at all. Using a slider will make the shared element slide away
+         // with the rest of of the content.
+         transitionHandler = rememberBackStackFader()
```
```suggestion
         transitionHandler = rememberBackStackFader(),
         withSharedElementTransition = true
```

manuelvicnt commented on libraries/core/src/main/kotlin/com/bumble/appyx/core/composable/Children.kt:
```diff
@@ -50,21 +59,49 @@ inline fun <reified NavTarget : Any, State> ParentNode<NavTarget>.Children(
+        ) {
+            CompositionLocalProvider(
+                /** If sharedElement is not supported for this Node - provide null otherwise children
+                 * can consume ascendant's  LocalSharedElementScope */
```
```suggestion
                 * can consume ascendant's LocalSharedElementScope */
```

manuelvicnt commented on libraries/core/src/main/kotlin/com/bumble/appyx/core/navigation/transition/MovableContent.kt:
```diff
@@ -0,0 +1,36 @@
+@Composable
+fun localMovableContentWithTargetVisibility(
+    key: Any,
+    defaultValue: @Composable () -> Unit
```
Nit: Instead of `defaultValue`, this should be `content` right?

manuelvicnt commented on libraries/core/src/main/kotlin/com/bumble/appyx/core/navigation/transition/MovableContent.kt:
```diff
@@ -0,0 +1,36 @@
+ * that becomes invisible to a Node that is becoming visible.
+ */
+@Composable
+fun localMovableContentWithTargetVisibility(
```
We might want to create multiple `localMovableContentWithTargetVisibility` functions with different `content` lambdas as there are many `movableContentOf` functions you can call? https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/runtime/runtime/src/commonMain/kotlin/androidx/compose/runtime/MovableContent.kt

KovalevAndrey replied:
Supported other functions too

manuelvicnt commented on libraries/core/src/main/kotlin/com/bumble/appyx/core/node/ParentNode.kt:
```diff
@@ -74,6 +76,15 @@ abstract class ParentNode<NavTarget : Any>(
+    @Composable
+    final override fun DerivedSetup(innerContent: @Composable () -> Unit) {
+        CompositionLocalProvider(
+            LocalParentNodeMovableContent provides mutableMapOf()
```
Should we create the `mutableMapOf` for every `ParentNode`? What if the `LocalParentNodeMovableContent` is set alongside `LocalSharedElementScope`? Is it possible? What are the implications?

This would effectively create a new map for every single `ParentNode` we have in the hierarchy, even if the Map isn't used

KovalevAndrey replied:
Moved to Children  composable and add it when movableContent is enabled

manuelvicnt commented on samples/sandbox/src/main/kotlin/com/bumble/appyx/sandbox/client/container/ContainerNode.kt:
```diff
@@ -96,6 +101,7 @@ class ContainerNode internal constructor(
         when (navTarget) {
             is Picker -> node(buildContext) { modifier -> ExamplesList(modifier) }
             is NavModelExamples -> NavModelExamplesNode(buildContext)
+            is SharedElementFaderExample -> SharedElementWithMovableContentExampleNode(buildContext)
```
Should we add the two examples? Shared element transitions with and without movable content

KovalevAndrey replied:
I can add another one yeah

manuelvicnt commented on samples/sandbox/src/main/kotlin/com/bumble/appyx/sandbox/client/sharedelement/FullScreenNode.kt:
```diff
@@ -0,0 +1,60 @@
+        ) {
+            val profile = Profile.allProfiles[profileId]
+
+
```
Nit: extra line

manuelvicnt reviewed (commented):
Looks good to me! Added some comments but overall looks great :D thank you!

KovalevAndrey committed:
Update documentation/ui/transitions.md

Co-authored-by: Manuel Vivo <manuelvicnt@gmail.com>

KovalevAndrey committed:
Update documentation/ui/transitions.md

Co-authored-by: Manuel Vivo <manuelvicnt@gmail.com>

KovalevAndrey committed:
Update libraries/core/src/main/kotlin/com/bumble/appyx/core/composable/Children.kt

Co-authored-by: Manuel Vivo <manuelvicnt@gmail.com>

KovalevAndrey committed:
Update documentation/ui/transitions.md

Co-authored-by: Manuel Vivo <manuelvicnt@gmail.com>

KovalevAndrey committed:
- Added shared element example without movable content
- Moved LocalMovableContentMap to Children composable and added a parameter to it.
- Added support for more movableContent API variants

manuelvicnt commented on documentation/ui/transitions.md:
````diff
@@ -12,9 +12,135 @@ Below you can find the different options how to visualise `NavModel` state chang
+
+```kotlin
+@Composable
+fun CustomMovableContent(key: Any, modifier: Modifier = Modifier) {
````
Add the `ParentNode` to this example so that people know where to set `withMovableContent = true`?

KovalevAndrey committed:
Add withMovableContent param to docs
